ADR/0018-apps-continuous-perf-testing.md

```diff
@@ -0,0 +1,72 @@
+
+What makes performance testing different from functional testing is it is harder to decide if the test passed or failed as every performance test tests different aspects of the application and so it expects different performance, different metrics, different thresholds. Furthermore, even if the measured performance of the app under test did not change, there might be some significant resource usage by the application which we want to cause the test to be marked as failed.
+
+The approach to make this pass/fail detection possible proposed here, is to use historical results as a benchmark – not only actual results of the performance test, but also monitoring data about the resource usage, etc.
+
+> Example: Imagine you develop a web e-shop application that uses PostgreSQL data backend. Your perf test is browsing through a goods catalog and measures latency of pages. When you want to decide on pass or fail result of this test, you need to check that metrics like below are aligned to previous results of a same test with same configuration:
+>
+> - Average page load latency.
+> - Backend service CPU and memory usage...
+> - PostgreSQL CPU and memory usage…
+> - Internal PostgreSQL metrics like number of commits during the test or average number of database sessions during the test…
+>
+> If any of these metrics does not align to historical results, we mark this test result as a failure.
+
+And even if some test fails, the application team should be able to review the data and change the test definition, so next time a new test result is being evaluated based on the historical results, this new result will be one of them.
+
+The algorithm that says if the current test passes or fails when compared to historical results can vary. It can be:
+
+- Checking minimum and maximum of the metric in historical data and making sure the new result is in this range.
+- Using a floating average of historical data and setting a safe range around it and making sure the new result is in that range.
+- Some machine learning thingy.
+
+Goal of this ADR is to propose a way for this kind of testing to be implemented in Stonesoup (feature [STONE-679](https://issues.redhat.com/browse/STONE-679)). Even if it would not deprecate full-blown performance and scale testing, having a release gate with some short and small scale performance test is desirable for many application teams.
+
+We are considering two different options here. They differ in what portion of logic mentioned above is executed where.
+
+### Glossary
+
+- SUT: Subject Under Test
+- Horreum: a results storage service: <https://github.com/Hyperfoil/Horreum>
+- CPT: Continuous Performance Testing
+
+## Decision
+
+Let's use this architecture with a single Horreum instance that would be managed by Red Hat and used by all customers.
```

Comment:
How resource intensive is the Horreum instance? Is it possible to separate out its use per workspace? @amisstea, FYI.

Reply:
Again, will ask John to answer here.

Reply:
There are other considerations wrt multiple instances vs a single instance. Horreum has been designed as a multi-tenanted system from the outset, with a comprehensive security model and Row Level Security implemented to prevent leaking of data between tenants. We could provide a "configuration" per workspace that would provide isolation between workspaces.  While it is possible to stand up and instance per workspace, that would require further development work (i.e. operators for provisioning per worksapce, data backup/restore etc)  and management would become more onerous on the SRE team(s)

Reply:
Hello @rvansa. Would you have some estimate on how resource intensive Horreum is please?

Reply:
@jhutar unfortunately @rvansa is no longer supporting Horreum. 

@arewm it is possible to spread the cost of running Horreum across multiple workspaces. wrt an actual figure of the cost to run the service, it depends on the SLA requirements and projected utilization levels. This something we are not aware of atm.

For our internal service, we host multiple teams atm using a single instance, we use 1 machine for the Horreum instance, 1 for postgres database and one for backup. 

We can provide details for resource usage for our current internal instance. We are also currently designing a scaling test to document how the resource requirements scale with load

With this information it should be possible to estimate the cost of running a Horreum service in a cloud provider